File: filly-core/src/session.rs
use anyhow::Result;
use crate::backend::Backend;
use crate::widget::{EventResult, Widget};

pub fn run(
    mut widget: impl Widget,
    backend: &mut impl Backend,
    tx: Option<std::sync::mpsc::Sender<serde_json::Value>>,
) -> Result<filly_protocol::WidgetResponse> {
    loop {
        if widget.is_dirty() {
            let area = backend.size();
            let tree = widget.render(area);
            backend.draw(&tree)?;
            widget.clear_dirty();
        }
        let event = backend.next_event()?;
        match widget.handle_event(&event) {
            EventResult::Response(resp) => return Ok(resp),
            EventResult::Yield(v) => {
                if let Some(ref tx) = tx { tx.send(v).ok(); }
            }
            EventResult::Handled => {}
            EventResult::Unhandled => {}
        }
    }
}

File: filly-core/session.c
#include <string.h>
#include <stdio.h>
#include "session.h"
#include "render.h"
#include <stdlib.h>

WidgetResponse session_run(Widget *w, Backend *backend) {
    if (!w || !backend || !backend->vtable) {
        WidgetResponse err = { .result = NULL, .cancelled = true, .error = "Invalid widget or backend" };
        return err;
    }
    if (!w->vtable.render || !w->vtable.handle_event ||
        !w->vtable.is_dirty || !w->vtable.clear_dirty) {
        WidgetResponse err = { .result = NULL, .cancelled = true, .error = "Incomplete widget vtable" };
        return err;
    }

    backend->vtable->setup(backend->data);
    int term_w, term_h;
    backend->vtable->get_size(backend->data, &term_w, &term_h);

    int idle_count = 0;

    while (1) {
        if (w->vtable.is_dirty(w)) {
            RenderTree tree;
            memset(&tree, 0, sizeof(tree));
            w->vtable.render(w, rect_new(0, 0, term_w, term_h), &tree);
            backend->vtable->draw(backend->data, &tree);
            render_tree_free(&tree);
            w->vtable.clear_dirty(w);
        }
        Event ev = backend->vtable->next_event(backend->data);
        if (ev.type == EVENT_NONE) {
            idle_count++;
            if (idle_count > 5000) {
                WidgetResponse timeout = { .result = NULL, .cancelled = false, .error = NULL };
                backend->vtable->teardown(backend->data);
                return timeout;
            }
            continue;
        }
        idle_count = 0;
        EventResult result = w->vtable.handle_event(w, &ev, backend);
        if (result.type == EVENT_RESULT_RESPONSE) {
            backend->vtable->teardown(backend->data);
            return result.response;
        }
    }
}